quinn-proto/src/connection/pacing.rs

```diff
@@ -0,0 +1,73 @@
+//! Pacing of packet transmissions.
+
+use std::{
+    convert::TryInto,
+    time::{Duration, Instant},
+};
+
+/// A simple token-bucket pacer. The bucket starts full and has an adjustable capacity. Once the
+/// bucket is empty, further transmission is blocked. The bucket refills at a rate determined by
+/// the congestion window and estimated round-trip time.
+pub struct Pacer {
+    capacity: u64,
+    tokens: u64,
+    last_instant: Instant,
+}
+
+impl Pacer {
+    /// Obtains a new [`Pacer`].
+    pub fn new(capacity: u64) -> Self {
+        Self {
+            capacity,
+            tokens: capacity,
+            last_instant: Instant::now(),
+        }
+    }
+
+    /// Return how long we need to wait before sending a packet.
+    ///
+    /// If we can send a packet right away, this returns [`None`]. Otherwise, returns `Some(d)`,
+    /// where `d` is the time before this function should be called again.
+    ///
+    /// We panic if the passed-in instant is older than `self.last_instant`.
```

Comment:
Since the `instant` comes from the user (indirectly), I'm not sure panicking is the right approach here. Is there some other way to respond here?

Reply:
I changed the code to ignore such timestamps.